Terminate idle resources

oliveigah opened the pull request:
https://github.com/dashbitco/nimble_pool/issues/1

Known problems with the implementantion:

  * On lazy pools, if no worker is currently on the pool the callback will never be called.
  Therefore you can not rely on this callback to terminate empty lazy pools.
  
  I'll make a follow up PR addressing this issue.

  * On not lazy pools, if you return `{:remove, user_reason}` you may end up
  terminating and initializing workers at the same time every idle verification cycle.
  
 I don't know exactly how to address this, probably will just add it to the docs. 

  * On large pools, if many resources goes idle at the same cycle you may end up terminating a large
  amount of workers sequencially, what could lead to the pool being unable to fulfill requests.
  
  I'll make a follow up PR addressing this issue, implementing something to stop the idle iterations earlier using some pool configuration probably.

josevalim commented on lib/nimble_pool.ex:
```diff
@@ -170,12 +170,37 @@ defmodule NimblePool do
               {:ok, pool_state}
 
+  @doc """
+  Handle pings due to inactivity on worker
+
+  Executed whenever the idle worker periodic timer verifies that a worker has been idle
+  on the pool for longer than `:resource idle_timeout` pool configuration milliseconds.
+
+  This callback must return one of the following values:
+    * `{:ok, worker_state}`: Updates worker state.
+
+    * `{:remove, user_reason}`: The pool will proceed to the standard worker termination
+        defined in `terminate_worker/3`.
+
+    * `{:stop, user_reason}`: The entire pool process will be terminated, and `terminate_worker/3`
+        will be called for every worker on the pool.
+
+    This callback is optional.
```
```suggestion
  Handle pings due to inactivity on worker.

  Executed whenever the idle worker periodic timer verifies that a worker has been idle
  on the pool for longer than `:resource_idle_timeout` pool configuration milliseconds.

  This callback must return one of the following values:

    * `{:ok, worker_state}` - Updates worker state.

    * `{:remove, user_reason}` - The pool will proceed to the standard worker termination
        defined in `terminate_worker/3`.

    * `{:stop, user_reason}` - The entire pool process will be terminated, and `terminate_worker/3`
        will be called for every worker on the pool.

  This callback is optional.
```

josevalim commented on lib/nimble_pool.ex:
```diff
@@ -170,12 +170,37 @@ defmodule NimblePool do
                       handle_enqueue: 2,
                       handle_update: 3,
-                      terminate_worker: 3
+                      terminate_worker: 3,
+                      handle_ping: 2
```
```suggestion
                      handle_ping: 2,
                      terminate_worker: 3
```

josevalim commented on lib/nimble_pool.ex:
```diff
@@ -213,11 +238,16 @@ defmodule NimblePool do
     * `:lazy` - When `true`, workers are started lazily, only when necessary.
       Defaults to `false`.
 
+    * `:resource_idle_timeout` - Timeout to tag a worker as idle.
```
Maybe we should rename this to `:worker_idle_timeout`? Especially given the worker nomenclature is the one we use throughout.

josevalim commented on lib/nimble_pool.ex:
```diff
@@ -481,14 +517,28 @@ defmodule NimblePool do
+  @impl true
+  def handle_info(:check_idle, %{resources: resources} = state) do
+    now_in_ms = System.monotonic_time(:millisecond)
+    resources_list = :queue.to_list(resources)
```
It shouldn't be necessary to convert the queue to a list. Instead take an item from the queue, see if it is idle, and continue. Otherwise, put it back on top.

oliveigah replied:
Yeah, I've chosen to turn it into a list to use `++` instead of `:queue.join/2` whenever just some workers are idle.

But looking at it now, I think it's not worth the parsing overhead.

josevalim commented on lib/nimble_pool.ex:
```diff
@@ -750,4 +868,10 @@ defmodule NimblePool do
+  defp get_metadata() do
+    %{
+      enqueued_at: System.monotonic_time(:millisecond)
+    }
```
No need for a map, just the monotonic time will be enough.

oliveigah replied:
I thought a map would be better here because it let the metadata attribute to be extended more easily with other fields beyond enqueued_time.

What you think?

josevalim replied:
If we ever need more metadata, then we can change it to a map. But right now it seems we only need the timestamp. :)

josevalim commented on lib/nimble_pool.ex:
```diff
@@ -330,11 +360,15 @@ defmodule NimblePool do
     lazy = if lazy, do: pool_size, else: nil
 
+    if resource_idle_timeout do
+      :timer.send_interval(resource_idle_timeout, :check_idle)
```
It is best to avoid a dependency on the `:timer` module, it reduces the chances of it becoming a bottleneck and of it also dogpilling (i.e. if the pool is slow we can have multiple check_idle's queued). Instead, use `Process.send_after(...)`.

josevalim replied:
If the `handle_ping` callback is not implemented, we should emit a warning here and not even schedule the interval.

oliveigah replied:
Good point! I did not catch that! ! :smile:

josevalim commented on test/nimble_pool_test.exs:
```diff
@@ -1344,4 +1352,139 @@ defmodule NimblePoolTest do
+          resource_idle_timeout: 5
+        )
+
+      :timer.sleep(6)
```
Instead of sleeping here, make it so handle_ping sends a message to the test process, and then you can do `assert_receive :pong`.

oliveigah replied:
Agreed with the `assert_receive` point.

But I did not get how this makes the sleep unnecessary. 

The sleep here is necessary because of the tests instructions approach.

Since the instructions are expected to be executed in order, I can call the next instruction only after the previous one is executed.

If the previous instruction is a `handle_ping` I can not trigger it manually, so I have to sleep for the timeout duration to guarantee the expected execution.

Did I get the tests instructions behaviour wrong? :thinking:

josevalim reviewed (commented):
Hi @oliveigah! This is looking great, I have submitted an initial review.

oliveigah commented:
@josevalim Thanks for the feedback!

Just finished the requested changes, let me know if you see some other way I can improve the feature.

Also, I've added a disclaimer section on the docs related to the known implementations issues on the PR description.

Gonna remove them on the follow up PRs.

ericmj commented on lib/nimble_pool.ex:
```diff
@@ -170,11 +170,49 @@ defmodule NimblePool do
+  terminating and initializing workers at the same time every idle verification cycle.
+
+  * On large pools, if many resources goes idle at the same cycle you may end up terminating a large
+  number of workers sequencially, what could lead to the pool being unable to fulfill requests.
```
Should this be indented like the list above?

ericmj commented on lib/nimble_pool.ex:
```diff
@@ -170,11 +170,49 @@ defmodule NimblePool do
+              worker_state,
+              pool_state
+            ) ::
+              {:ok, worker_state} | {:remove, user_reason} | {:stop, user_reason()}
```
```suggestion
              {:ok, worker_state} | {:remove, user_reason()} | {:stop, user_reason()}
```

ericmj commented on lib/nimble_pool.ex:
```diff
@@ -330,11 +373,21 @@ defmodule NimblePool do
     lazy = if lazy, do: pool_size, else: nil
 
+    if worker_idle_timeout do
+      if function_exported?(worker, :handle_ping, 2) do
```
Is this check really needed? If you add the options you should be expected to implement the callback.

oliveigah replied:
Do you have any suggestion about how can we enforce both being provided together?

Since I did not find a way to enforce it, I think a warning would be a good option beyond did not start the timer in the first place.

ericmj replied:
I would assume that we should just crash if `handle_ping/2` is not implemented. But that might not be the best solution. I defer to @josevalim.

xinz replied:
In my opinion, the added L381-L390 lines put into `GenServer.handle_continue/2` would be better to ensure the `:check_idle` operation will be executed after the process's initialization is finished.

josevalim replied:
@ericmj the issue is that projects like Finch may expose all of NimblePool options to the user, so a user may accidentally set this and break their pool. Having a warning upfront leads to a better UX.

@xinz this check is very fast, so there is no reason to move it to handle_continue.

xinz replied:
Hello @josevalim, thanks, yes, it should be fast as expect, but since it will apply workers' `init_worker/1` callback later, there maybe some use cases would take more time to initialize, do we need to take more for these cases?

josevalim replied:
It is up to the worker to decide if they want eager or lazy initialization, the library should allow both. :)

oliveigah commented:
@josevalim @ericmj @xinz 

Thank you for the feedback! I've just added a commit that tackles the problem:

> On large pools, if many resources goes idle at the same cycle you may end up terminating a large
> amount of workers sequentially, what could lead to the pool being unable to fulfill requests.

I've added a pool configuration option called `:max_idle_pings` which defines a limit to the number of workers that can be pinged for each cycle of the `handle_ping/2` optional callback. Defaults to `nil`, which is no limit.

There are two behaviours I think it is important to keep in mind:

- If you are not terminating workers with handle_ping/2, you may end up pinging only the same workers over and over again because each cycle will ping only the first :max_idle_pings workers

- If you are terminating workers with handle_ping/2, the last worker may be terminated after up to `worker_idle_timeout + worker_idle_timeout * ceil(number_of_workers/max_idle_pings)` instead of `2 * worker_idle_timeout - 1` milliseconds of idle time.

Let me know what you think! And if you think about a better approach. :smiley:

josevalim commented on lib/nimble_pool.ex:
```diff
@@ -170,11 +170,65 @@ defmodule NimblePool do
+    * `{:stop, user_reason}`: The entire pool process will be terminated, and `terminate_worker/3`
+        will be called for every worker on the pool.
+
+    This callback is optional.
```
```suggestion
  This callback is optional.
```

josevalim commented on lib/nimble_pool.ex:
```diff
@@ -170,11 +170,65 @@ defmodule NimblePool do
+  of workers. But it is important to keep in mind the following behaviours whenever utilizing it.
+
+    * If you are not terminating workers with `handle_ping/2`, you may end up pinging only the same
+    workers over and over again because each cycle will ping only the first `:max_idle_pings` workers
```
```suggestion
      workers over and over again because each cycle will ping only the first `:max_idle_pings` workers
```

josevalim commented on lib/nimble_pool.ex:
```diff
@@ -170,11 +170,65 @@ defmodule NimblePool do
+    workers over and over again because each cycle will ping only the first `:max_idle_pings` workers
+
+    * If you are terminating workers with `handle_ping/2`, the last worker may be terminated after up to
+    `worker_idle_timeout + worker_idle_timeout * ceil(number_of_workers/max_idle_pings)`
+     instead of `2 * worker_idle_timeout - 1` milliseconds of idle time.
```
```suggestion
      `worker_idle_timeout + worker_idle_timeout * ceil(number_of_workers/max_idle_pings)`
       instead of `2 * worker_idle_timeout - 1` milliseconds of idle time.
```

josevalim replied:
Let's also add an example with actual values. Something like: "For instance, if you have a pool with 10 resources, and a ping of 1 second, without max_idle_pings, the last one will be terminated in X. If you have max_idle_pings of 2, it will take X seconds. A max_idle_pings of 5 (half), will take Y seconds."

josevalim commented on lib/nimble_pool.ex:
```diff
@@ -170,11 +170,65 @@ defmodule NimblePool do
+    a large number of workers sequencially, what could lead to the pool being unable to
+    fulfill requests. See `:max_idle_pings` option to prevent this.
+
+
+
+
```
```suggestion
```

josevalim commented on lib/nimble_pool.ex:
```diff
@@ -170,11 +170,65 @@ defmodule NimblePool do
+  ## Disclaimers
+
+    * On lazy pools, if no worker is currently on the pool the callback will never be called.
+    Therefore you can not rely on this callback to terminate empty lazy pools.
```
```suggestion
      Therefore you can not rely on this callback to terminate empty lazy pools.
```

josevalim commented on lib/nimble_pool.ex:
```diff
@@ -170,11 +170,65 @@ defmodule NimblePool do
+    Therefore you can not rely on this callback to terminate empty lazy pools.
+
+    * On not lazy pools, if you return `{:remove, user_reason}` you may end up
+    terminating and initializing workers at the same time every idle verification cycle.
```
```suggestion
      terminating and initializing workers at the same time every idle verification cycle.
```

josevalim commented on lib/nimble_pool.ex:
```diff
@@ -170,11 +170,65 @@ defmodule NimblePool do
+    terminating and initializing workers at the same time every idle verification cycle.
+
+    * On large pools, if many resources goes idle at the same cycle you may end up terminating
+    a large number of workers sequencially, what could lead to the pool being unable to
+    fulfill requests. See `:max_idle_pings` option to prevent this.
```
```suggestion
      a large number of workers sequentially, what could lead to the pool being unable to
      fulfill requests. See `:max_idle_pings` option to prevent this.
```

josevalim commented on lib/nimble_pool.ex:
```diff
@@ -213,11 +267,21 @@ defmodule NimblePool do
+      Defaults to `nil`
+
+    * `:max_idle_pings` - Defines a limit to the number of workers that can be pinged
+    for each cycle of the `handle_ping/2` optional callback.
+    Defaults to `nil`, which is no limit. See `handle_ping/2` for more details.
```
```suggestion
      for each cycle of the `handle_ping/2` optional callback.
      Defaults to `nil`, which is no limit. See `handle_ping/2` for more details.
```

josevalim reviewed (changes requested):
Initial pass on the docs. More feedback coming soon, sorry for the delay!

josevalim commented on lib/nimble_pool.ex:
```diff
@@ -330,11 +398,21 @@ defmodule NimblePool do
+        Process.send_after(self(), :check_idle, worker_idle_timeout)
+      else
+        IO.warn(
+          "worker_idle_timeout is not nil but handle_ping/2 callback is not exported by the worker."
```
```suggestion
          ":worker_idle_timeout was given but the worker does not export a handle_ping/2 callback"
```

josevalim commented on lib/nimble_pool.ex:
```diff
@@ -646,6 +742,90 @@ defmodule NimblePool do
+
+      {{:value, resource_data}, resources} ->
+        {worker_server_state, worker_metadata} = resource_data
+        time_diff = abs(worker_metadata - now_in_ms)
```
```suggestion
        time_diff = now_in_ms - worker_metadata
```

josevalim commented on lib/nimble_pool.ex:
```diff
@@ -646,6 +742,90 @@ defmodule NimblePool do
+    do_check_idle_resources(resources, now_in_ms, state, :queue.new(), 0)
+  end
+
+  defp do_check_idle_resources(resources, now_in_ms, state, new_resources, pinged_workers) do
```
I think you should handle the `max_idle_pings` here. Set `max_idle_pings` to default to -1 (if none is given). Then you call this function as:

```elixir
do_check_idle_resources(resources, now_in_ms, state, :queue.new(), max_idle_pings)
```

Then you add a clause for:

```
do_check_idle_resources(resources, now_in_ms, state, new_resources, 0)
```

That's when you know you can stop pinging. :)

oliveigah replied:
Nice, I think it is clearer this way too!

josevalim commented on lib/nimble_pool.ex:
```diff
@@ -646,6 +742,90 @@ defmodule NimblePool do
+  defp maybe_ping_worker(worker_server_state, state) do
+    %{worker: worker, state: pool_state} = state
+
+    if function_exported?(worker, :handle_ping, 2) do
```
This check is not necessary. At this point, handle_ping/2 should always be exported?

oliveigah replied:
Yes, you are correct it is a left over from before implementing your suggestion to not start the timer at all if it is not implemented!

oliveigah commented:
@josevalim Thanks for the feedback. Just finished the suggested improvements.

Let me know if you see some other way to make it better!

josevalim commented on lib/nimble_pool.ex:
```diff
@@ -213,11 +273,21 @@ defmodule NimblePool do
+      Defaults to `nil`
+
+    * `:max_idle_pings` - Defines a limit to the number of workers that can be pinged
+      for each cycle of the `handle_ping/2` optional callback.
+      Defaults to -1, which means limit. See `handle_ping/2` for more details.
```
```suggestion
      for each cycle of the `handle_ping/2` optional callback.
      Defaults to no limit. See `handle_ping/2` for more details.

```

josevalim commented on lib/nimble_pool.ex:
```diff
@@ -213,11 +273,21 @@ defmodule NimblePool do
+    * `:worker_idle_timeout` - Timeout in milliseconds to tag a worker as idle.
+      If not nil, starts a periodic timer on the same frequency that will ping
+      all idle workers using `handle_ping/2` optional callback .
+      Defaults to `nil`
```
```suggestion
      Defaults to no timeout.
```

josevalim commented:
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:
